Add check to shortcodes

corywebb opened the pull request:
# Description

Add checks in shortcodes for whether or not the Fast checkout buttons should be hidden.

# Testing instructions

1. Add a shortcode: `[fast_cart /]` or `[fast_product /]`
2. Verify that the button shows when it should and is hidden when it should be.

# Checks
- [x] Code has been formatted with `phpcbf --standard WordPress`
- [x] No new linter warnings from `phpcs --standard WordPress`

@ilkerulutas @brikr @ejcho623 this is ready for a review.

brikr commented on includes/shortcodes.php:
```diff
@@ -31,10 +31,16 @@ function fastwc_add_shortcodes() {
-		'buttons/fast-checkout-button',
-		$atts
-	);
+	if ( fastwc_should_hide_checkout_button() ) {
```
FWIW, this function does rely on two global variables that are set in the `wc_product_addon_start` and `woocommerce_grouped_product_list_before` hooks. Depending on where they try to jam this shortcode, those variables might not be set. It also relies on the global `$product` variable.
I think it's probably fine to keep this as is, but it's something we should keep in the back of our heads in case it comes up in a corner case.

corywebb replied:
@brikr I'm revisiting this. Do you know why we use those globals to check for addons and grouped products? There is another check for addons in the cart button check, and I'm wondering why we don't just use that to check if the product has addons - https://github.com/fast-af/fast-checkout-woocommerce/blob/main/includes/checkout.php#L158-L167.

I'm sure we can find a similar way to check for grouped products as well, and just remove the dependency on the hooks and global variables. What do you think?

brikr reviewed (approved):
one fyi comment but otherwise lgtm

corywebb commented:
@brikr I reworked the hide button logic for both the PDP and cart buttons, so it should be more easily extendable in the future. I also updated the logic for determining if a product is an addon product or a grouped product, and that should now work across all scenarios.

Can you re-review this when you get a chance?

brikr commented on includes/checkout.php:
```diff
@@ -43,284 +46,51 @@ function fastwc_get_cart_data() {
  */
 function fastwc_woocommerce_after_add_to_cart_quantity() {
-	if ( fastwc_should_hide_checkout_button() ) {
+	if ( fastwc_should_hide_pdp_checkout_button() ) {
 		return;
 	}
 
 	fastwc_load_template( 'fast-pdp' );
```
should we have a `maybe_render_pdp_button` function that we use here that is similar to the one you wrote for cart button?

corywebb replied:
I added the `maybe_render_cart_button` because the same logic is used in multiple places, but for the sake of consistency I think it's a good idea to add the `maybe render_pdp_button` function as well. I'll get that added.

corywebb replied:
I updated the function to be a more general `fastwc_maybe_render_checkout_button` and added a `$button_type` parameter.

brikr commented on includes/hide.php:
```diff
@@ -0,0 +1,290 @@
+ * @return bool
+ */
+function fastwc_should_hide_pdp_checkout_button( $product_id = 0 ) {
+	if ( empty( $product_id ) && is_product() ) {
+		global $product;
+	} elseif ( ! empty( $product_id ) ) {
+		$product = wc_get_product( $product_id );
+	}
```
nice, i really like this

brikr commented on includes/hide.php:
```diff
@@ -0,0 +1,290 @@
+add_filter( 'fastwc_should_hide_pdp_checkout_button', 'fastwc_should_hide_pdp_button_for_out_of_stock_product', 4, 2 );
+
+/**
+ * Determine if the Fast PDP button shoudl be hidden for an external product (i.e. not purchased on this store).
```
```suggestion
 * Determine if the Fast PDP button should be hidden for an external product (i.e. not purchased on this store).
```

corywebb replied:
Good catch.

brikr reviewed (changes requested):
real small change. i like these refactors, nice work!

corywebb commented:
@brikr this is now updated per your change request.

brikr reviewed (approved):
i dig it. nice work!
